contracts/DomainRegistry.sol

```diff
@@ -0,0 +1,114 @@
+        require(keccak256(abi.encodePacked(parentDomain)) == keccak256(abi.encodePacked("")) || domains[parentDomain].controller != address(0), "Parent domain must exist");
+
+        if (keccak256(abi.encodePacked(parentDomain)) != keccak256(abi.encodePacked("")) && domains[parentDomain].controller != address(0)) {
+            domains[parentDomain].childDomains.push(keccak256(abi.encodePacked(cleanDomainName)));
+        }
+
+        domains[cleanDomainName] = Domain({
+            controller: msg.sender,
+            deposit: msg.value,
+            childDomains: new bytes32[](0)
+        });
+    }
+
+    function releaseDomain(string memory domainName) public onlyController(domainName) {
+        string memory cleanDomainName = DomainValidator.stripProtocol(domainName);
+        require(domains[cleanDomainName].childDomains.length == 0,"Remove child domains first");
+        require(domains[cleanDomainName].controller != address(0),"Domain is not reserved");
+
+        removeChildFromParentDomain(cleanDomainName);
+
+        address payable controller = payable(msg.sender);
+        uint256 refundAmount = domains[cleanDomainName].deposit;
+        delete domains[cleanDomainName];
+
+        controller.transfer(refundAmount);
+    }
+
+    function removeChildFromParentDomain(string memory childDomain) private {
+        string memory parentDomain = DomainValidator.getParentDomain(childDomain);
+        bytes32 hashedChild = keccak256(abi.encodePacked(childDomain));
+        bytes32 hashedParent = keccak256(abi.encodePacked(parentDomain));
+
+        if (hashedParent != keccak256(abi.encodePacked("")) &&domains[parentDomain].controller != address(0)) {
+            bytes32[] storage childDomains = domains[parentDomain].childDomains;
+            uint index;
+            bool found = false;
+
+            for (uint i = 0; i < childDomains.length; i++) {
+                if (childDomains[i] == hashedChild) {
+                    index = i;
+                    found = true;
+                    break;
+                }
+            }
+
+            if (found) {
+                assembly {
+                    let len := sload(childDomains.slot)
+                    if lt(index, sub(len, 1)) {
+                        let data := add(add(childDomains.slot, 1), index)
+                        let lastElem := add(
+                            add(childDomains.slot, 1),
+                            sub(len, 1)
+                        )
+                        sstore(data, sload(lastElem))
+                    }
+                    sstore(childDomains.slot, sub(len, 1))
+                }
+            }
+        }
+    }
+
+    function changeDeposit(string memory domainName, uint256 newDeposit) public payable onlyController(domainName) sufficientDeposit(newDeposit) {
+        string memory cleanDomainName = DomainValidator.stripProtocol(domainName);
+        domains[cleanDomainName].deposit = msg.value;
+    }
```

Comment:
Two comments here:
1) There is no need to over complicate the system. It was neither in implicit or explicit requirements that the deposit for the domain should change. Besides that, what's the value of this in general?
2) There is a bug. What if you reseved domain with 1 enter, then changed deposit to 2 ether. That means that you'll get back only 2 ether but not 3 when releasing it.

        -2 points